File: android/src/main/java/com/android/xknowledge/framework/aspectj/login/CheckLoginAspectJ.java
package com.android.xknowledge.framework.aspectj.login;

import android.content.Context;
import android.util.Log;
import android.widget.Toast;

import com.android.xknowledge.XApplication;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.aspectj.lang.reflect.MethodSignature;

@Aspect
public class CheckLoginAspectJ {
    private static final String TAG = "CheckLogin";

    /**
     * 找到处理的切点
     * * *(..)  可以处理CheckLogin这个类所有的方法
     */
    @Pointcut("execution(@com.android.xknowledge.framework.aspectj.login.CheckLogin  * *(..))")
    public void executionCheckLogin() {
    }

    /**
     * 处理切面
     *
     * @param joinPoint
     * @return
     */
    @Around("executionCheckLogin()")
    public Object checkLogin(ProceedingJoinPoint joinPoint) throws Throwable {
        Log.i(TAG, "checkLogin: ");
        MethodSignature signature = (MethodSignature) joinPoint.getSignature();
        CheckLogin checkLogin = signature.getMethod().getAnnotation(CheckLogin.class);
        if (checkLogin != null) {
            Context context = (Context) joinPoint.getThis();
            if (XApplication.Companion.isLogin()) {
                Log.i(TAG, "checkLogin: 登录成功 ");
                return joinPoint.proceed();
            } else {
                Log.i(TAG, "checkLogin: 请登录");
                Toast.makeText(context, "请登录", Toast.LENGTH_SHORT).show();
                return null;
            }
        }
        return joinPoint.proceed();
    }
}

File: android/src/main/java/com/android/xknowledge/framework/aspectj/login/CheckLoginAspectJ.kt
package com.android.xknowledge.framework.aspectj.login

import android.content.Context
import android.util.Log
import android.widget.Toast
import com.android.xknowledge.XApplication
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.aspectj.lang.reflect.MethodSignature

@Aspect
class CheckLoginAspectJ {
    private val TAG = "CheckLogin"

    @Pointcut("com.com.android.xknowledge.framework.aspectj.login.CheckLogin  * *(..))")
    fun executionCheckLogin() {
    }

    @Around("executionCheckLogin()")
    @Throws(Throwable::class)
    fun checkLogin(joinPoint: ProceedingJoinPoint): Any? {
        Log.i(TAG, "checkLogin: ");
        val signature = joinPoint.signature as MethodSignature
        val checkLogin = signature.method.getAnnotation(CheckLogin::class.java)
        if (checkLogin != null) {
            val context = joinPoint.getThis() as Context
            return if (XApplication.isLogin) {
                joinPoint.proceed()
            } else {
                Toast.makeText(context, "请登录", Toast.LENGTH_SHORT).show()
                null
            }
        }
        return null
    }
}